src/encoder/text.rs

```diff
@@ -224,26 +218,49 @@ fn label_pairs_to_text(
 
 /// `escape_string` replaces `\` by `\\`, new line character by `\n`, and `"` by `\"` if
 /// `include_double_quote` is true.
-fn escape_string(v: &str, include_double_quote: bool) -> String {
-    let mut escaped = String::with_capacity(v.len() * 2);
+///
+/// Implementation adapted from
+/// https://lise-henry.github.io/articles/optimising_strings.html
+fn escape_string(v: &str, include_double_quote: bool) -> Cow<'_, str> {
+    // Regex compilation is expensive. Use `lazy_static` to compile the regexes
+    // once per process lifetime and not once per function invocation.
+    lazy_static! {
+        static ref ESCAPER: Regex = Regex::new("(\\\\|\n)").expect("Regex to be valid.");
+        static ref QUOTED_ESCAPER: Regex = Regex::new("(\\\\|\n|\")").expect("Regex to be valid.");
+    }
 
-    for c in v.chars() {
-        match c {
-            '\\' | '\n' => {
-                escaped.extend(c.escape_default());
-            }
-            '"' if include_double_quote => {
-                escaped.extend(c.escape_default());
-            }
-            _ => {
-                escaped.push(c);
+    let first_occurence = if include_double_quote {
+        QUOTED_ESCAPER.find(v)
+    } else {
+        ESCAPER.find(v)
+    }
+    .as_ref()
+    .map(Match::start);
+
+    if let Some(first) = first_occurence {
+        let mut escaped = String::with_capacity(v.len() * 2);
+        escaped.push_str(&v[0..first]);
+        let remainder = v[first..].chars();
+
+        for c in remainder {
+            match c {
+                '\\' | '\n' => {
+                    escaped.extend(c.escape_default());
+                }
+                '"' if include_double_quote => {
+                    escaped.extend(c.escape_default());
+                }
+                _ => {
+                    escaped.push(c);
+                }
             }
         }
+        escaped.into()
+    } else {
+        // The input string does not contain any characters that would need to
+        // be escaped. Return it as it is.
+        v.into()
     }
-
-    escaped.shrink_to_fit();
```

Comment:
This is still relevant for the `Some(first)` arm, right?

Reply:
Ah, good catch. Sorry for missing that. Speeds up the escaping test a bit:

```bash
$ cargo benchcmp optimized optimized_shrink                                   
 name                                 optimized ns/iter  optimized_shrink ns/iter  diff ns/iter  diff %  speedup    
 bench_text_encoder_with_escaping     105,675,740        99,516,087                  -6,159,653  -5.83%   x 1.06    
 bench_text_encoder_without_escaping  73,070,167         72,932,821                    -137,346  -0.19%   x 1.00
```
Fix included in https://github.com/tikv/rust-prometheus/pull/327/commits/93b35e23a658de3888501c6da3b1712e5137b94e.